lib/net.c

```diff
@@ -50,32 +51,231 @@ find_num_numa_nodes(void)
+		rte_eth_dev_close(iface->ports[i]);
 	}
 }
 
+static void
+close_iface_id(struct gatekeeper_if *iface, uint8_t nb_slave_ports)
+{
+	uint8_t i;
+
+	/* If there's only one port, there's no bonded port. */
+	if (iface->num_ports == 1)
+		return;
+
+	for (i = 0; i < nb_slave_ports; i--) {
+		rte_eth_bond_slave_remove(iface->id, iface->ports[i]);
+	}
+
+	rte_eth_bond_free(iface->name);
+}
+
+static void
+close_iface(struct gatekeeper_if *iface)
+{
+	close_iface_id(iface, iface->num_ports);
+	close_iface_ports(iface, iface->num_ports);
+	free(iface->ports);
+}
+
 struct net_config *
 get_net_conf(void)
 {
 	return &config;
 }
 
+static int
+init_port(struct net_config *net_conf, uint8_t port_id, uint8_t *num_succ_ports)
+{
+	int ret;
+	uint32_t num_lcores;
+	uint32_t lcore;
+	struct rte_eth_link link;
+
+	ret = rte_eth_dev_configure(port_id,
+		net_conf->num_rx_queues, net_conf->num_tx_queues,
+		&gatekeeper_port_conf);
+	if (ret < 0) {
+		RTE_LOG(ERR, PORT,
+			"Failed to configure port %hhu (err=%d)!\n",
+			port_id, ret);
+		return ret;
+	}
+
+	num_lcores = rte_lcore_count();
+	for (lcore = 0; lcore < num_lcores; lcore++) {
+		size_t numa_node;
+		
+		/* XXX Map queue = lcore, if necessary,
+		 * change the queues mapping.
+		 */
+		uint16_t queue = (uint16_t)lcore;
+
+		/* XXX In case the number of lcores is greater than
+		 * the number of queues.
+		 */
+		if (lcore >= net_conf->num_rx_queues ||
+				lcore >= net_conf->num_tx_queues)
+			break;
+
+		numa_node = rte_lcore_to_socket_id(lcore);
+
+		ret = rte_eth_rx_queue_setup(port_id, queue,
+			GATEKEEPER_NUM_RX_DESC, (unsigned int)numa_node, NULL,
+			net_conf->gatekeeper_pktmbuf_pool[numa_node]);
+		if (ret < 0) {
+			RTE_LOG(ERR, PORT, "Failed to configure port %hhu rx_queue %hu (err=%d)!\n",
+				 port_id, queue, ret);
+			return ret;
+		}
+
+		ret = rte_eth_tx_queue_setup(port_id, queue,
+			GATEKEEPER_NUM_TX_DESC, numa_node, NULL);
+		if (ret < 0) {
+			RTE_LOG(ERR, PORT, "Failed to configure port %hhu tx_queue %hu (err=%d)!\n",
+				 port_id, queue, ret);
+			return ret;
+		}
+	}
+
+	/* Start device. */
+	ret = rte_eth_dev_start(port_id);
+	if (ret < 0) {
+		RTE_LOG(ERR, PORT, "Failed to start port %hhu (err=%d)!\n",
+			port_id, ret);
+		return ret;
+	}
+	if (num_succ_ports != NULL)
+		(*num_succ_ports)++;
+
+	/*
+	 * The following code ensures that the device is ready for
+	 * full speed RX/TX.
+	 * When the initialization is done without this,
+	 * the initial packet transmission may be blocked.
+	 */
+	rte_eth_link_get(port_id, &link);
+	if (!link.link_status) {
+		RTE_LOG(ERR, PORT, "Querying port %hhu, and link is down!\n",
+			port_id);
+		ret = -1;
+		return ret;
+	}
+
+	/* TODO Configure the Flow Director and RSS. */
+
+	return 0;
+}
+
+static int
+init_iface(struct net_config *net_conf, struct gatekeeper_if *iface)
+{
+	int ret = 0;
+	uint8_t i;
+	uint8_t num_succ_ports = 0;
+	uint8_t num_slaves_added = 0;
+
+	iface->ports = malloc(sizeof(*iface->ports) * iface->num_ports);
```

Comment:
Should we use ``rte_malloc()`` or ``rte_zmalloc()`` instead of ``malloc()`` or ``calloc()`` for all the cases? These functions may perform better in NUMA systems.